mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -141,12 +144,14 @@ impl PeerManager {
         // TODO: What about garbage collection of validators
         if peer.raw_network_id().is_validator_network() {
             if let Some(state) = self.peer_states.lock().get_mut(&peer) {
+                counters::active_upstream_peers(&peer.raw_network_id()).dec();
                 state.is_alive = false;
             }
         } else {
             // All other nodes have their state immediately restarted anyways, so let's free them
             // TODO: Why is the Validator optimization not applied here
             self.peer_states.lock().remove(&peer);
+            counters::active_upstream_peers(&peer.raw_network_id()).dec();
```

Comment:
nit: I still think this logic is a little weird (i.e., as called out by the TODO). For example, I'd only decrement the counter if it's not in the peer_states (even in this case), to protect against any weird bugs where we get a removal notification for a peer but we never got a new peer notice, but this shouldn't happen. Either way, we should probably clean all this stuff up at some point 😄  For now, it's probably good enough!

Reply:
Yeah, I don't get it either, but I didn't have time to figure out why it does it really...